<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.543">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>deseq2-summary</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="main_files/libs/clipboard/clipboard.min.js"></script>
<script src="main_files/libs/quarto-html/quarto.js"></script>
<script src="main_files/libs/quarto-html/popper.min.js"></script>
<script src="main_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="main_files/libs/quarto-html/anchor.min.js"></script>
<link href="main_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="main_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="main_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="main_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="main_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="main_files/libs/quarto-dashboard/quarto-dashboard.js"></script>
<script src="main_files/libs/quarto-dashboard/stickythead.js"></script>
<script src="main_files/libs/quarto-dashboard/datatables.min.js" kdttablesentinel="true"></script>
<script src="main_files/libs/quarto-dashboard/pdfmake.min.js" kdttablesentinel="true"></script>
<script src="main_files/libs/quarto-dashboard/vfs_fonts.js" kdttablesentinel="true"></script>
<script src="main_files/libs/quarto-dashboard/web-components.js" type="module"></script>
<script src="main_files/libs/quarto-dashboard/components.js"></script>
<link href="main_files/libs/quarto-dashboard/datatables.min.css" rel="stylesheet" kdttablesentinel="true">


</head>

<body class="quarto-dashboard dashboard-fill dashboard-sidebar fullcontent">

<header id="quarto-dashboard-header">
<nav class="navbar navbar-expand-md slim" data-bs-theme="dark">
  <div class="navbar-container container-fluid">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#dashboard-collapse" aria-controls="dashboard-collapse" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>    


    <div class="navbar-brand-container">       

      <div class="navbar-title">
        <div class="navbar-title-text"><a href="#">deseq2-summary</a></div>
        
        
      </div>
    </div>
<div id="dashboard-collapse" class="navbar-collapse collapse"><ul class="navbar-nav navbar-nav-scroll me-auto" role="tablist"><li class="nav-item" role="presentation"><a id="tab-results-summary" class="nav-link active" data-bs-toggle="tab" role="tab" data-bs-target="#results-summary" data-scrolling="false" href="#results-summary" aria-controls="results-summary" aria-selected="true"><span class="nav-link-text">Results Summary</span></a></li><li class="nav-item" role="presentation"><a id="tab-pca-explorer" class="nav-link" data-bs-toggle="tab" role="tab" data-bs-target="#pca-explorer" data-scrolling="false" href="#pca-explorer" aria-controls="pca-explorer" aria-selected="false"><span class="nav-link-text">PCA-Explorer</span></a></li></ul></div></div>
</nav>
</header>
<div class="page-layout-custom quarto-dashboard-content bslib-gap-spacing html-fill-container bslib-page-fill quarto-dashboard-pages dashboard-sidebar-container">  

<div class="bslib-sidebar-layout html-fill-item html-fill-container" data-bslib-sidebar-open="desktop" data-bslib-sidebar-init="true" style=" --bslib-sidebar-width: 300px" data-bslib-sidebar-border="false" data-bslib-sidebar-border-radius="false"><div class="main html-fill-container html-fill-item"><div class="sidebar-content html-fill-item html-fill-container">
<div class="tab-content bslib-grid html-fill-item html-fill-container" style="display: grid; grid-template-columns: minmax(3em, 1fr);
grid-auto-rows: minmax(0, 1fr);">
<div id="results-summary" class="dashboard-page tab-pane show active html-fill-item html-fill-container" data-orientation="rows" aria-labelledby="tab-results-summary">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 15fr) minmax(3em, 85fr); grid-auto-columns: minmax(0, 1fr);">
<!-- #------------------------------------------------------->
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-columns: minmax(3em, 1fr) minmax(3em, 1fr) minmax(3em, 1fr);
grid-auto-rows: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()">
<div class="card-body html-fill-item html-fill-container" style="padding: 0;">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="valbox1" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div><div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()">
<div class="card-body html-fill-item html-fill-container" style="padding: 0;">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="valbox2" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div><div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()">
<div class="card-body html-fill-item html-fill-container" style="padding: 0;">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="valbox3" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
<div class="card tabset html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()">
<div class="card-header dashboard-card-no-title bslib-navs-card-title">

<ul class="nav nav-tabs card-header-tabs" role="tablist" data-tabsetid="card-tabset-4"><li class="nav-item" role="presentation"><a href="#card-tabset-4-1" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="DESeq2 Results" aria-selected="true" class="nav-link active">DESeq2 Results</a></li><li class="nav-item" role="presentation"><a href="#card-tabset-4-2" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="Normalized Counts" aria-selected="false" class="nav-link">Normalized Counts</a></li><li class="nav-item" role="presentation"><a href="#card-tabset-4-3" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="Density Distributions" aria-selected="false" class="nav-link">Density Distributions</a></li><li class="nav-item" role="presentation"><a href="#card-tabset-4-4" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="MA-Plot" aria-selected="false" class="nav-link">MA-Plot</a></li></ul></div>




<div class="tab-content html-fill-item html-fill-container" data-tabset-id="card-tabset-4"><div class="tab-pane active show html-fill-item html-fill-container" role="tabpanel" id="card-tabset-4-1"><div class="card-body html-fill-item html-fill-container" data-title="DESeq2 Results">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 1fr); grid-auto-columns: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-body html-fill-item html-fill-container">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="uploaded" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div></div><div class="tab-pane html-fill-item html-fill-container" role="tabpanel" id="card-tabset-4-2"><div class="card-body html-fill-item html-fill-container" data-title="Normalized Counts">
<div class="html-fill-item html-fill-container dashboard-toolbar-container"><div class="toolbar toolbar-bottom html-fill-container" data-layout="flow">
<div class="cell-output-display no-overflow-x">
<div id="csv" class="shiny-html-output"></div>
</div>
</div><div class="toolbar-content html-fill-item html-fill-container">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 1fr); grid-auto-columns: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-body html-fill-item html-fill-container">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="spinnerSpark" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div></div>
</div></div><div class="tab-pane html-fill-item html-fill-container" role="tabpanel" id="card-tabset-4-3"><div class="card-body html-fill-item html-fill-container" data-title="Density Distributions">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 1fr); grid-auto-columns: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-body html-fill-item html-fill-container">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="DENS" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div></div><div class="tab-pane html-fill-item html-fill-container" role="tabpanel" id="card-tabset-4-4"><div class="card-body html-fill-item html-fill-container" data-title="MA-Plot">
<div class="html-fill-item html-fill-container dashboard-toolbar-container"><div class="toolbar toolbar-bottom html-fill-container" data-layout="flow">
<div class="cell-output-display no-overflow-x">
<div id="ma.button" class="shiny-html-output"></div>
</div>
</div><div class="toolbar-content html-fill-item html-fill-container">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 1fr); grid-auto-columns: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-body html-fill-item html-fill-container">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="MA" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div></div>
</div></div></div><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div><div id="pca-explorer" class="dashboard-page tab-pane grid-skip html-fill-item html-fill-container" data-orientation="rows" aria-labelledby="tab-pca-explorer">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-rows: minmax(3em, 1fr); grid-auto-columns: minmax(0, 1fr);">
<div class="card tabset html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()">
<div class="card-header dashboard-card-no-title bslib-navs-card-title">

<ul class="nav nav-tabs card-header-tabs" role="tablist" data-tabsetid="card-tabset-9"><li class="nav-item" role="presentation"><a href="#card-tabset-9-1" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="PCA Plot" aria-selected="true" class="nav-link active">PCA Plot</a></li><li class="nav-item" role="presentation"><a href="#card-tabset-9-2" role="tab" data-toggle="tab" data-bs-toggle="tab" data-value="PCA Variance" aria-selected="false" class="nav-link">PCA Variance</a></li></ul></div>


<div class="tab-content html-fill-item html-fill-container" data-tabset-id="card-tabset-9"><div class="tab-pane active show html-fill-item html-fill-container" role="tabpanel" id="card-tabset-9-1"><div class="card-body html-fill-item html-fill-container dashboard-sidebar-container" data-title="PCA Plot">
<div class="bslib-sidebar-layout html-fill-item sidebar-right html-fill-container" data-bslib-sidebar-open="desktop" data-bslib-sidebar-init="true" style=" --bslib-sidebar-width: 300px"><div class="main html-fill-container html-fill-item"><div class="sidebar-content html-fill-item html-fill-container">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-columns: minmax(3em, 60fr);
grid-auto-rows: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-body html-fill-item html-fill-container" style="padding: 10px;">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="PCA3" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div></div><aside id="bslib-sidebar-2" class="sidebar html-fill-container html-fill-item"><div class="html-fill-item html-fill-container">
<div class="cell-output-display no-overflow-x">
<div data-spinner-id="spinner-da297adc3d2a4c142065a6c878cd63d6" class="shiny-spinner-output-container shiny-spinner-hideui">
<div class="load-container shiny-spinner-hidden load1">
<div id="spinner-da297adc3d2a4c142065a6c878cd63d6" class="loader">Loading...</div>
</div>
<div style="height:400px" class="shiny-spinner-placeholder"></div>
<div id="metaID" class="shiny-html-output"></div>
</div>
</div>
<div class="cell-output-display no-overflow-x">
<div id="pcx" class="shiny-html-output"></div>
</div>
<div class="cell-output-display no-overflow-x">
<div id="pcy" class="shiny-html-output"></div>
</div>
<div class="cell-output-display no-overflow-x">
<div id="pcz" class="shiny-html-output"></div>
</div>
</div></aside>
<button class="collapse-toggle" type="button" title="Toggle sidebar" aria-expanded="true" aria-controls="bslib-sidebar-2">
    <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 16 16" class="bi bi-chevron-left collapse-icon" style="fill:currentColor;" aria-hidden="true" role="img">
      <path fill-rule="evenodd" d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"></path>
    </svg>
</button>
<script data-bslib-sidebar-init="">
bslib.Sidebar.initCollapsibleAll()
</script>
</div>
</div></div><div class="tab-pane html-fill-item html-fill-container" role="tabpanel" id="card-tabset-9-2"><div class="card-body html-fill-item html-fill-container" data-title="PCA Variance">
<div class="html-fill-item html-fill-container bslib-grid" style="display: grid; grid-template-columns: minmax(3em, 1fr);
grid-auto-rows: minmax(0, 1fr);">
<div class="card cell html-fill-item html-fill-container bslib-card" data-title="Top 100 Genes Contributing to PC1" data-bslib-card-init="" data-require-bs-caller="card()" data-full-screen="false">
<div class="card-header"><div class="card-title html-fill-item html-fill-container">Top 100 Genes Contributing to PC1</div></div>
<div class="card-body html-fill-item html-fill-container" style="padding: 10px;">
<div class="cell-output-display html-fill-item html-fill-container no-overflow-x">
<div id="pcaLoadings" class="shiny-html-output html-fill-item html-fill-container"></div>
</div>
</div>
<bslib-tooltip placement="auto" bsoptions="[]" data-require-bs-version="5" data-require-bs-caller="tooltip()">
    <template>Expand</template>
    <span class="bslib-full-screen-enter badge rounded-pill">
        <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 24 24" style="height:1em;width:1em;" aria-hidden="true" role="img"><path d="M20 5C20 4.4 19.6 4 19 4H13C12.4 4 12 3.6 12 3C12 2.4 12.4 2 13 2H21C21.6 2 22 2.4 22 3V11C22 11.6 21.6 12 21 12C20.4 12 20 11.6 20 11V5ZM4 19C4 19.6 4.4 20 5 20H11C11.6 20 12 20.4 12 21C12 21.6 11.6 22 11 22H3C2.4 22 2 21.6 2 21V13C2 12.4 2.4 12 3 12C3.6 12 4 12.4 4 13V19Z"></path></svg>
    </span>
</bslib-tooltip><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div><div id="3ade8a4a-fb1d-4a6c-8409-ac45482d5fc9" class="hidden html-fill-item html-fill-container">

</div>
</div>
</div></div></div><script data-bslib-card-init="">bslib.Card.initializeAllCards();</script></div>
</div>
</div>
</div>
</div></div><aside id="bslib-sidebar-1" class="sidebar html-fill-container html-fill-item"><div class="html-fill-container html-fill-item">
<div class="cell-output-display">
<div class="form-group shiny-input-container" style="width:100%;">
<label class="control-label" id="file-label" for="file">Import DESeq2-Final Results</label>
<div class="input-group">
<label class="input-group-btn input-group-prepend">
<span class="btn btn-default btn-file">
Browse...
<input id="file" class="shiny-input-file" name="file" type="file" style="position: absolute !important; top: -99999px !important; left: -99999px !important;" accept=".txt">
</span>
</label>
<input type="text" class="form-control" placeholder="No file selected" readonly="readonly">
</div>
<div id="file_progress" class="progress active shiny-file-input-progress">
<div class="progress-bar"></div>
</div>
</div>
</div>
<div class="cell-output-display">
<button id="run" type="button" class="btn btn-default action-button">Submit</button>
</div>
<div class="cell-output-display">
<hr>
</div>
<div class="cell-output-display no-overflow-x">
<div id="contrastID" class="shiny-html-output"></div>
</div>
<div class="cell-output-display">
<hr>
</div>
<div class="cell-output-display">
<br>
</div>
<div class="cell-output-display no-overflow-x">
<div id="filtCounts" class="shiny-html-output"></div>
</div>
<div class="cell-output-display">
<br>
</div>
<div class="cell-output-display">
<br>
</div>
<div class="cell-output-display no-overflow-x">
<div id="mat.switch" class="shiny-html-output"></div>
</div>
<div class="cell-output-display no-overflow-x">
<div id="fractionT" class="shiny-html-output"></div>
</div>
</div></aside>
<button class="collapse-toggle" type="button" title="Toggle sidebar" aria-expanded="true" aria-controls="bslib-sidebar-1">
    <svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 16 16" class="bi bi-chevron-left collapse-icon" style="fill:currentColor;" aria-hidden="true" role="img">
      <path fill-rule="evenodd" d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"></path>
    </svg>
</button>
<script data-bslib-sidebar-init="">
bslib.Sidebar.initCollapsibleAll()
</script>
</div>
<p>
<script type="application/shiny-prerendered" data-context="server-start">
library(quarto)
library(shiny)
library(dplyr)
library(tibble)
library(tidyverse)
library(shinycssloaders)
library(heatmaply)
library(shinyjs)
library(shiny.fluent)
library(colourpicker)
library(shiny.blueprint)
library(shinyjqui)
library(tidyr)
library(sparkline)
library(ggplot2)
library(scales)
library(ggpubr)
library(shinyWidgets)
library(bslib)
library(flexdashboard)
library(bsicons)
library(DT)
library(reactable)
library(reshape2)
library(reactablefmtr)
library(htmltools)
library(htmlwidgets)
library(tippy)


fileInput("file", label = "Import DESeq2-Final Results", accept = ".txt", width = "100%")
actionButton(inputId="run","Submit")
hr()
uiOutput(outputId = "contrastID")
hr()
br()
uiOutput(outputId = "filtCounts")

br()
br()

uiOutput(outputId = "mat.switch")
uiOutput("fractionT")

targetFile <- read.delim("targetFile.txt", header = T, sep = "\t")
targetFile$norm <- paste0("norm.", targetFile$label)
targetFile$group <- as.factor(targetFile$group)


pca_rds <- readRDS("/Users/fa286/Desktop/5081D/STAR.COUNTS/rawCounts/5081D/5081D_PCA_Matrix_nTop_500.RDS")

# Assosiate Group Levels with Sample_ID ====
fmain = function(group_){
  
  x = list(targetFile %>% filter(group == group_) %>% pull(norm))
  y = group_
  
  names(x) <- paste0("Group: ", y)
  
  return(x)
}


l5 <- list() 
  
for (i in 1:length(levels(targetFile$group))) {

  l5[[i]] <- fmain(group_ = levels(targetFile$group)[i])
    
}

l6 <- lapply(l5, function(x){
  
 x = noquote(paste0("colGroup(name = '", names(x), "', columns =", noquote(paste0(x)) ,")"))
  
})

</script>
 
<script type="application/shiny-prerendered" data-context="server">

pca.df <- pca_rds$PCA.df

pca_rotation <- pca_rds$prcomp.out$rotation

loading_scores <- pca_rotation [,1:5]
ab.loading_scores <- abs(loading_scores)



output$loadings <- renderReactable(reactable(data.frame(ab.loading_scores) %>% mutate_if(is.numeric, round, 3), 
                          resizable = T, striped = T, searchable = T, filterable = F, 
                          defaultPageSize = 50, showPageInfo = F, bordered = T, highlight = T,compact = F, 
                          defaultSorted = c("PC1"), showPageSizeOptions = T,
                          showSortable = T,paginationType = "simple",
                          columns = list(PC1 = colDef(defaultSortOrder = "desc", sortable = T)
                          ),
                          width = 1200, defaultColDef = colDef(minWidth = 150, maxWidth = 500, align = "left"), 
                          theme = 
                            reactableTheme(
                              borderColor    = "#c7c7c7",
                              stripedColor   = "#f6f8fa",
                              highlightColor = "#eddcd8",
                              cellPadding = "8px 12px",
                              style = list(
                                fontFamily = "-apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif"
                              ),
                              searchInputStyle = list(width = "100%"))
                          
                          
))


output$pcaLoadings <- renderUI(reactableOutput(outputId = "loadings"))

 # render-ui -- cluster  ---------------------------------------------------------------
  
  output$metaID <- renderUI({
    
    
    tagList(
      selectInput(inputId = "metaclusterID", label = "Select Metadata",
                  choices = colnames(pca.df %>% select(!matches("PC"))), selected = "group")
      
    )
    
  })
  
  

  
  # render-ui -- pcx --------------------------------------------------------
  
  output$pcx <- renderUI({
    
    u_sel_meta = req(input$metaclusterID)
    
    tagList(
      selectInput(inputId = "pcx_ID", label = "Select PC of x-axis",
                  choices = colnames(pca.df %>% select(matches("PC"))), selected = "PC1"))
    
    
  })
  
  # render-ui -- pcy --------------------------------------------------------
  
  output$pcy <- renderUI({
    
    u_sel_meta = req(input$metaclusterID)
    
    tagList(
      selectInput(inputId = "pcy_ID", label = "Select PC of y-axis",
                  choices = colnames(pca.df %>% select(matches("PC"))), selected = "PC2"))
    
    
  })
  
  
  # render-ui -- pcz --------------------------------------------------------
  
  output$pcz <- renderUI({
    
    u_sel_meta = req(input$metaclusterID)
    
    tagList(
      selectInput(inputId = "pcz_ID", label = "Select PC of z-axis",
                  choices = colnames(pca.df %>% select(matches("PC"))), selected = "PC3"))
    
    
  })
  







  
output$pca3d <- 
    
    renderPlotly({
      
      u_sel_pcx  =  req(input$pcx_ID)
      u_sel_pcy = req(input$pcy_ID)
      u_sel_pcz = req(input$pcz_ID)
      
      # --------------------------------------------------------------------------------------------------------

      m <- list(
        l = 1,
        r = 1,
        b = 50,
        t = 50,
        pad = 1
      )
      plot_ly(data = pca.df,  x = ~  get(u_sel_pcx), y= ~ get(u_sel_pcy), z = ~  get(u_sel_pcz), 
              width = 1800, height = 1200,
              color = ~ get(input$metaclusterID),
              colors = c('#636EFA','#EF553B','#00CC96'),
              text = pca.df$label,
              marker = list(size = 14,
                            line = list(color = ~  ~ get(input$metaclusterID) , width = 1)),
              hovertemplate = paste("<b>%{text}<b><br> PCx: %{x}<br> PCy: %{y}<br> PCz: %{z}")) %>%
        add_markers() %>%
        layout(autosize = T, margin =m, title = "3D Principal Components Analysis",
               scene = list(xaxis = list(title = paste0(input$pcx_ID, ": ", (pca_rds$Variance.df %>% filter(x == input$pcx_ID) %>% pull(percentVar)), " %")),
                            yaxis = list(title = paste0(input$pcy_ID, ": ", (pca_rds$Variance.df %>% filter(x == input$pcy_ID) %>% pull(percentVar)), " %")),
                            zaxis = list(title = paste0(input$pcz_ID, ": ", (pca_rds$Variance.df %>% filter(x == input$pcz_ID) %>% pull(percentVar)), " %"))
               ), margin = m
        ) %>% layout(legend = list(orientation = 'v', xanchor = "left"))
      
      
      
      
      
      
      
    })

output$PCA3 <- renderUI(withSpinner((plotlyOutput(outputId = "pca3d", width = 800, height = 800))))














db <- eventReactive(input$run, {
  if (!is.null(input$file$datapath)) {
    read.delim(input$file$datapath, header = T, sep = "\t") 
  } else {
    stop()
  }
})


# Reactive DropDown Menu in Sidebar -- Allows user to choose available DESeq2 Contrasts + All_Comparisons ====
observeEvent(input$run , {
  
  output$contrastID <- renderUI({

    tagList(
        tags$style(type='text/css', ".selectize-dropdown-content {height: 1000px; }"),
        h5("Select DESeq2 Comparison"),
        selectInput(inputId = "pID", label = "", width = 600,
                    choices = c("All_Comparisons", sapply((strsplit(colnames(db())[grep("*.log2FoldChange", x = colnames(db()))], split = "\\.")), 
                                     FUN = function(x){purrr::pluck(x, 1)})),  
                    selected = "All_Comparisons")
      )
    })
  
  }
  
  
)


# Reactive function to parse the nume and denom for a given contrast ====

con.num <- reactive({

  req(input$pID)

  if (input$pID != "All_Comparisons") {

    a = input$pID

    a1 = str_split(a, pattern = "_vs_")[[1]][1]
    a2 = str_split(a, pattern = "_vs_")[[1]][2]
    #

    return(a1)
    
    }

})



con.denom <- reactive({

  req(input$pID)

  if (input$pID != "All_Comparisons") {

    a = input$pID

    a1 = str_split(a, pattern = "_vs_")[[1]][1]
    a2 = str_split(a, pattern = "_vs_")[[1]][2]
    #

    return(a2)
    
    }

})


# Data Subset Reactive Objects:

# 1. Normalized Counts: ====

norm.counts = reactive({
  req(db())
  db() %>% select(gene_id, gene, baseMean, matches("norm")) %>% tibble::column_to_rownames("gene_id") })

# 2. Filtered Norm Counts Based on basemean quantiles: ====

quants <- reactive({
  quantile(db()$baseMean)
})

norm.counts.filtered = reactive({
  
  req(input$min.counts)
  x = norm.counts()
  x %>% filter(baseMean >= quants()[unname(paste0(input$min.counts, "%"))])

  
  
  
})

# 3. SparkLines Setup: ====

normCounts.tab <- reactive({
  
  req(norm.counts.filtered())

  m1 = norm.counts.filtered()  %>% rownames_to_column("GENE") 
  
  m2 = m1 %>% select(-baseMean, -gene) %>% melt()

  m3 = m2 %>% group_by(GENE) %>% summarise(Trend = list(value))
  
  m4 = left_join(m1, m3, by = c("GENE")) %>% column_to_rownames("GENE")
 
  return(m4)
  
})

# 4. db2 -- comparison selector:====

observeEvent(input$pID,

             if (input$pID =="All_Comparisons") {
               output$mat.switch <- renderUI({tagList(
                      hr(),
                      #materialSwitch(inputId = "switch2",label = "", value = FALSE, status = "warning")
                      )})
             } else if (input$pID != "All_Comparisons") {
                output$mat.switch <- renderUI({tagList(
                      p("Please use this toggle button to switch the numerator and denominator for the selected comparison. "),
                      materialSwitch(inputId = "switch2",label = "Swap Numerator/Denominator", value = FALSE, status = "warning"),
                      p("For the selected comparison:"), 
                      # p(paste0("Numerator Group = ", strsplit(input$pID, split = "_")[[1]][1])),
                      # p(paste0("Denominator/Baseline Group = ", strsplit(input$pID, split = "_")[[1]][3])),   
                      # 
                      # p(paste0("Numerator Group = ")),
                      # p(paste0("Denominator/Baseline Group = "))
                    

                      
                      
                      )
                  
                      })
             }

)

# 4.1 db2 ====
db2 <- reactive({
  req(input$file$datapath)
  req(input$pID)
  req(db())
  
  
  if (input$pID == "All_Comparisons") {
    db() %>% rowwise() %>% mutate(n_comparisons_sig = sum(c_across(matches("padj")) <= 0.05, na.rm = T)) %>%
      select(!matches("norm")) %>% filter(gene_id %in% rownames(norm.counts.filtered())) %>% 
      mutate(Significance = ifelse(n_comparisons_sig == 0, "ns", "Yes")) %>% select(-n_comparisons_sig) %>% 
      select(Significance, everything())
    
  } else if (input$pID !="All_Comparisons"){
    
    a = db() %>% 
          select(gene_id, gene, locus, strand, gene_biotype, matches(input$pID)) 

    b = a %>% rowwise() %>% mutate(n_comparisons_sig = sum(c_across(matches("padj")) <= 0.05, na.rm = T)) %>%
          filter(gene_id %in% rownames(norm.counts.filtered())) %>%
          mutate(Significance = ifelse(n_comparisons_sig == 0, "ns", "Yes")) %>% select(-n_comparisons_sig) %>% 
          select(Significance, everything())
    
    b

  }
  
})

# 4.1.2 db3 ====
db3 <- reactive({
  
  req(db2())
  
  if (is.null(input$switch2)) {
    
    db2()
  
    } else if (input$switch2 == TRUE){
    
    db2() %>% filter(gene_id %in% rownames(norm.counts.filtered())) %>% mutate(across(matches("log2FoldChange"), ~ . * -1))
      
    }  else {
    
      db2()
  }
  
  
})

with_tooltip <- function(value, tooltip, ...) {
  div(style = "text-decoration: underline; cursor: help",
      tippy(value, tooltip, ...))
}

output$uploadedDb3 <- renderReactable(reactable(db3() %>% mutate_if(is.numeric, round, 3), 
                                                resizable = T, striped = T, searchable = T, filterable = F, 
          defaultPageSize = 50, showPageInfo = T, bordered = T, highlight = T,compact = T,
          columns = list(
            gene = colDef( filterable = T),
            strand = colDef(minWidth = 100),
            
            Significance = colDef(minWidth = 100,
              cell = function(value) {
                    if (value == "Yes") {
                      classes <- "tag num-high"
                    } else if (value == "ns") {
                      classes <- "tag num-med"
                    } else {
                      classes <- "tag num-low"
                    }
                    value <- format(value, nsmall = 1)
                    span(class = classes, value)
              }, header = with_tooltip("Significance", "When All_Comparisons is selected, 
                                       Significance Column displays Yes 
                                       where a given row/gene is significant in at least 1 comparison")
            )
        
            
          ),
          width = 1920, defaultColDef = colDef(minWidth = 200, maxWidth = 500, align = "center"), 
          theme = 
            reactableTheme(
                    borderColor    = "#c7c7c7",
                    stripedColor   = "#f6f8fa",
                    highlightColor = "#eddcd8",
                    cellPadding = "8px 12px",
                    style = list(
                      fontFamily = "-apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif"
                    ),
                    searchInputStyle = list(width = "100%"))
          
          
          ))

output$uploaded <- renderUI(withSpinner(reactableOutput(outputId = "uploadedDb3"), type = 1))



# output$uploaded <- DT::renderDataTable(DT::datatable(db3() %>% mutate_if(is.numeric, round, 3),
#                                                      list(paging=T,pageLength = 100, scrollX=T,scrollY=900,
#                                                           autoWidth = F), filter = "none",
#                                                      rownames = T, caption = "DESeq2 Results"))
# 


# Reactive Switch Behavior Num/Denom ====
observeEvent(input$switch2,{
  if (input$switch2 == TRUE && input$pID != "All_Comparisons" && !is.null(input$switch2)) {
    
    frac <- reactive({
     req(con.num())  
     a = c("Numerator", "Denominator/Baseline")
     b = c(con.denom(), con.num())
     data.frame(a,b)
    })
    
    output$fractionT <- renderUI(reactableOutput(outputId = "myT"))

    output$myT <- renderReactable({
    reactable(frac(), searchable = F, columns = list(a = colDef(maxWidth = 150, name = ""), 
                                                     b = colDef(maxWidth = 150, name = "", 
                                                              style = function(value){list(background = "#c9ada7", fontWeight = "bold")},
                                                              cell = function(value){span(class = "myCLASS1", value)})),
                                          columnGroups = list(colGroup(name = "log2FC", columns = c("a", "b"))),
                                          sortable = F, striped = F, outlined = T , resizable = T, bordered = F, highlight = T)
  })
    
  } else if(input$switch2 == FALSE && input$pID != "All_Comparisons" &&  !is.null(input$switch2)){
    
    frac <- reactive({
     req(con.num())  
     a = c("Numerator", "Denominator/Baseline")
     b = c(con.num(), con.denom())
     data.frame(a,b)
    })
    
    output$fractionT <- renderUI(reactableOutput(outputId = "myT"))

    output$myT <- renderReactable({
    reactable(frac(), searchable = F, columns = list(a = colDef(maxWidth = 150, name = ""), 
                                                     b = colDef(maxWidth = 150, name = "",
                                                    style = function(value){list(background = "#C7C7C7", fontWeight = "bold")})),
              columnGroups = list(colGroup(name = "log2FC", columns = c("a", "b"))),
                                                     sortable = T, resizable = T, bordered = F, striped = T, highlight = T)
  })
  }
})








# output$nCountsDT <- DT::renderDT(DT::datatable(normCounts.tab() %>% select(Trend, everything()), list(paging=T,
#                                                                              pageLength = 100, 
#                                                                              scrollX=T, 
#                                                                              scrollY=900,
#                                                                              scrollCollapse = TRUE,
#                                                                              autoWidth = F, dom = 'lftip',
#                                                                              fnDrawCallback = htmlwidgets::JS(
#                                                                               '
#                                                                               function(){
#                                                                                 HTMLWidgets.staticRender();
#                                                                               }
#                                                                               ')), 
#                                                       filter = "none", rownames = T, 
#                                                       caption = "DESeq2 Normalized Counts", 
#                                                       escape = F) %>% spk_add_deps() ) 

# output$nCountsDT <- renderReactable(reactable(normCounts.tab() %>% select(Trend, everything()),
#           columns = list(
#             Trend = colDef(cell =react_sparkline(normCounts.tab(), 
#                                                  line_curve = "linear",
#                                                  decimals = 1,
#                                                  show_area = T,
                                                
#                                                  highlight_points = highlight_points(first="orange",last="blue")), minWidth = 100)
#           ), 
#           resizable = T, striped = F, searchable = T, filterable = F, 
#           defaultPageSize = 50, showPageInfo = T, bordered = T, highlight = T,compact = T,
#           columnGroups = lapply(l6, function(x){eval(parse(text = x))})
# ))




# html_code <- '
#   setInterval(function(){
#     $(".inlinesparkline").sparkline();
#     $(".inlinesparkline").toggleClass("inlinesparkline noinlinesparkline");
#   }, 50);
# '

# Reactive Normalized Counts Table with Sparklines ====
tbl <- reactive({ 
  
  req(normCounts.tab())
  
  tb <-
  reactable(normCounts.tab() %>% select(Trend, everything()),
          columns = list(
            Trend = colDef(
              html = TRUE,
              cell = function(value, index){
                sparkline(normCounts.tab()$Trend[[index]], width = 70, height = 20)
              }, align = "center", maxWidth = 100
            
                )
            
          ), 
          resizable = T, striped = T, searchable = T, filterable = F, 
          defaultPageSize = 50, showPageInfo = T, bordered = T, highlight = T,compact = T,
          columnGroups = lapply(l6, function(x){eval(parse(text = x))}),
          width = 1920, 
          theme = 
            reactableTheme(
                    borderColor = "#c7c7c7",
                    stripedColor = "#f6f8fa",
                    highlightColor = "#eddcd8",
                    cellPadding = "8px 12px",
                    style = list(
                      fontFamily = "-apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif"
                    ),
                    searchInputStyle = list(width = "100%"))
) 
  # adding a custom JavaScript function to render the sparklines from the html tags
  # once the sparkline is created, the class is changed from inlinesparkline to noinlinesparkline
  # changing the class is necessary because otherwise another rendering would remove the created sparkoines
  # this function is run each 20ms so that it is applied even to sparklines on other pages of the reactable

#   normCounts.tab() %>%
#   rowwise() %>%
#   mutate(
#     # preparing the sparklines html manually
#     # sparklines html looks like this <span class="inlinesparkline">1,2,3,4,5<\u002fspan>
#     Trend = Trend %>%
#       unlist() %>%
#       paste(collapse = ",") %>%
#       paste0('<span class="inlinesparkline">', ., '<\u002fspan>')
#   ) %>%
#   reactable(
#     columns = list(
#       Trend = colDef(html = T) # rendering the content as html
#     ), 
#     columnGroups = lapply(l6, function(x){eval(parse(text = x))}), borderless = F, bordered = T, striped = F, resizable = T,compact = T,
#     highlight = T,height = 1240, outlined = T,
#     defaultColDef = colDef(minWidth = 10)
  # ) %>%
  # # adding a custom JavaScript function to render the sparklines from the html tags
  # # once the sparkline is created, the class is changed from inlinesparkline to noinlinesparkline
  # # changing the class is necessary because otherwise another rendering would remove the created sparkoines
  # # this function is run each 20ms so that it is applied even to sparklines on other pages of the reactable
  # prependContent(onStaticRenderComplete(
  #   HTML("setInterval(function(){ 
  #           $('.inlinesparkline').sparkline();
  #           $('.inlinesparkline').toggleClass('inlinesparkline noinlinesparkline');
  #         }, 50);
  #        ")
  # ))


# tb$dependencies <- getDependency("sparklines") 
tb

})



output$nCountsDT <- renderReactable({ 
  tbl()
  })




# Reactive Caption for ValBox 3 ====
cap <- reactive({
    req(input$pID)
    req(db2())
    if (input$pID == "All_Comparisons" && !is.null(input$pID)) {
      
      cap = paste0(length(sapply((strsplit(colnames(db())[grep("*.log2FoldChange", x = colnames(db()))], split = "\\.")), 
                                     FUN = function(x){purrr::pluck(x, 1)})), ";DESeq2 contrasts available in this experiment")
    } else {
    
      cap = paste0(length(which(db2() %>% select(matches(".padj")) < 0.05)), ";Differentially Expressed Features Found at padj < 0.05 for ", input$pID)
  }
  
  cap
  
})



# Value Boxes ====
output$valbox1 <- renderUI({
  validate(need(input$file, message = "  
    Please Upload DESeq2 results distributed by TREx and Click Submit!"))
  validate(need(db(), label = "Checks for the presence of db reactable"))
  value_box(title = "Total Annotated Features", value = textOutput(outputId = "v1"), 
            showcase = bs_icon("receipt"), 
            theme = value_box_theme(bg = "#4a4e69"), 
            class = "text-light", showcase_layout = "top right")
})

output$valbox2 <- renderUI({
  validate(need(input$file, message = ""))
  validate(need(db(), label = "Checks for the presence of db reactable"))
  total.filtered <- nrow(norm.counts.filtered())

  value_box(title = paste0("Features remaining with ", input$min.counts, "% quantile filtering" ),
            value =total.filtered, 
            showcase = bs_icon("funnel"), 
            theme = value_box_theme(bg = "#9a8c98"), 
            class = "text-light", showcase_layout = "top right")
})


output$valbox3 <- renderUI({
  validate(need(input$file, message = ""))
  validate(need(db(), label = "Checks for the presence of db reactable"))
  value_box(title =  strsplit(cap(), split = ";")[[1]][2], 
            value = strsplit(cap(), split = ";")[[1]][1], icon = "fa-dna", color = "#c9ada7", caption = strsplit(cap(), split = ";")[[1]][2], 
            showcase = bs_icon("arrow-down-up"), 
            theme = value_box_theme(bg = "#c9ada7"), 
            class = "text-light", showcase_layout = "top right")
})

output$v1 <- renderText(nrow(db()))
# output$v2 <- renderText(nrow(db()))
# output$v3 <- renderText(nrow(db()))





# output$norm.counts.raw <- DT::renderDataTable(DT::datatable(norm.counts() %>% mutate_if(is.numeric, round, 3), 
#                                                      list(paging=T,pageLength = 100, scrollX=T,scrollY=900, 
#                                                           autoWidth = F), filter = "none", 
#                                                      rownames = T, caption = "DESeq2 Normalized Counts")) 
# 
# output$norm.counts.filtered <- DT::renderDataTable(DT::datatable(norm.counts.filtered() %>% mutate_if(is.numeric, round, 3), 
#                                                      list(paging=T,pageLength = 100, scrollX=T,scrollY=900, 
#                                                           autoWidth = F), filter = "none", 
#                                                      rownames = T, caption = "DESeq2 Normalized Filtered Counts")) 

# Slider Filter for baseMean filtering on NormCounts + ggdensity vertical ====
observeEvent(input$run, {

  if (!is.null(input$file$datapath)) {
      output$filtCounts <- renderUI({
    
    tagList(
      tags$style(HTML(".js-irs-0 .irs-single, .js-irs-0 .irs-bar-edge, .js-irs-0 .irs-bar {background: #4a4e69}")),
       h5("Filter Counts"),
       p("Use the slider to filter rows i.e. genes/features based on quantiles of baseMean."),
       sliderInput(inputId = "min.counts", label = "Select min quantile", min = 0, max = 75,
                   value =  25, width = "75%", step = 25),
       hr()
      )
    })
  } else{
    return(NULL)
  }
  }
)

# ui-normalized-counts-w/sparkline ====
output$spinnerSpark <- renderUI(withSpinner(reactableOutput(outputId = "nCountsDT"), type = 1))


norm.stack <- reactive({
  x <- stack(norm.counts() %>% select(-baseMean,-gene))
  colnames(x) <- c("Counts", "Sample")
  return(x)
})

# DensityPlot ====
output$dist.plot <- 
    renderPlotly(
      ggplotly(ggdensity(norm.stack(), x = "Counts",
          rug = F,
          color = "Sample", fill = "Sample", alpha = 0.1, title = "Density Distribution of Normalized Counts") + 
            scale_x_log10() + theme_gray()  + geom_vline(xintercept= input$min.counts, linetype = "dashed") +
            theme(legend.position = "right"), tooltip = c("colour"))
          )

output$DENS <- renderUI(withSpinner(jqui_resizable((plotlyOutput(outputId = "dist.plot", width = 1000, height = 800)))))


# MA-Plot ====
maplot <- reactive({
  validate(need(input$pID !="All_Comparisons", message = "Please Select a DESeq2 Comparison", label = "---"))
  req(input$pID)
  
  if (input$pID != "All_Comparisons" && !is.null(input$switch2) && input$switch2 == FALSE) {
    main = input$pID
    maplot.data = db() %>% select(gene_id, baseMean, matches(input$pID))
    colnames(maplot.data) <-  gsub(pattern = paste0(input$pID,"."), replacement = "", x = colnames((maplot.data)))
    return(list(main = main, maplot.data = maplot.data))
    
    #return(maplot.data)
  } else if (input$pID != "All_Comparisons" && !is.null(input$switch2) && input$switch2 == TRUE){
    
    main = input$pID
    mainA = strsplit(input$pID, split = "_vs_")[[1]][1]
    mainB = strsplit(input$pID, split = "_vs_")[[1]][2]
    main2 = paste0(mainB, "_vs_", mainA)
    
    maplot.data = db() %>% select(gene_id, baseMean, matches(input$pID))
    colnames(maplot.data) <-  gsub(pattern = paste0(input$pID,"."), replacement = "", x = colnames((maplot.data)))
    maplot.data = maplot.data %>% mutate(across(matches("log2FoldChange"), ~ . * -1))
    return(list(main = main2, maplot.data = maplot.data))
  }
  
  
  
  
})


ma <- reactive({

    req(maplot())
    req(input$pID)

      ggmaplot(maplot()[[2]] , main = maplot()[[1]],
         fdr = 0.05, fc = 2, size = 4,
         palette = c("#B31B21", "#1465AC", "darkgray"),
         genenames = as.vector(maplot()[[2]]$gene_id),
         legend = "top", top = 20,
         font.label = c("bold", 14),
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) +
    theme(
    text = element_text(size = 16),  # Adjust the font size here
    plot.title = element_text(size = 20, face = "bold"),  # Customize title font size
    axis.title = element_text(size = 18),  # Customize axis title font size
    axis.text = element_text(size = 18)  # Customize axis label font size
)

})

output$ma.plot <- renderPlot({
    ma()
})

output$downloadCSV <- downloadHandler(
    filename = function() {
      paste0("Quantile_", input$min.counts, "_Filtered_Normalized_Counts_", Sys.Date(), ".csv", sep="")
    },
    content = function(file) {
      write.csv(normCounts.tab() %>% select(-Trend), file)
    }
    
)

observeEvent(!is.null(norm.counts.filtered()), {
  
  output$csv <- renderUI(downloadButton("downloadCSV", "Download Filtered Normalized Counts", style="width:100%;"))
})



output$maDownload <- downloadHandler(
    filename = function() {
      paste0("MA-Plot_", input$pID, "_", Sys.Date(), ".png", sep="")
    },
    content = function(file) {
      ggsave(file, ma(), bg = "white", width = 10, height = 10, units = "in", dpi = "print")
    }
    
)

# ui-maDownload ===
observeEvent(!is.null(db()), {
  output$ma.button <- renderUI(downloadButton("maDownload", "Download MA-Plot", style="width:100%;"))
})


# ui-MAPlot ====
output$MA <- renderUI(withSpinner(jqui_resizable(jqui_draggable(plotOutput(outputId = "ma.plot", width = 1080, height = 1080)))))




</script>
 
<script type="application/shiny-prerendered" data-context="server-extras">
ojs_define <- function(..., .session = shiny::getDefaultReactiveDomain()) {
  quos <- rlang::enquos(...)
  vars <- rlang::list2(...)
  nm <- names(vars)
  if (is.null(nm)) {
    nm <- rep_len("", length(vars))
  }
  mapply(function(q, nm, val) {
    # Infer name, if possible
    if (nm == "") {
      tryCatch({
        nm <- rlang::as_name(q)
      }, error = function(e) {
        code <- paste(collapse = "\n", deparse(rlang::f_rhs(q)))
        stop("ojs_define() could not create a name for the argument: ", code)
      })
    }
    .session$output[[nm]] <- val
    outputOptions(.session$output, nm, suspendWhenHidden = FALSE)
    .session$sendCustomMessage("ojs-export", list(name = nm))
    NULL
  }, quos, nm, vars, SIMPLIFY = FALSE, USE.NAMES = FALSE)
  invisible()
}
</script>
</p>
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["shinycssloaders-binding"]},{"type":"character","attributes":{},"value":["1.0.0.9011"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["assets"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["spinner.js"]},{"type":"character","attributes":{},"value":["spinner.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["shinycssloaders"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.0.0.9011"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113]}},"value":[{"type":"character","attributes":{},"value":["abind","assertthat","backports","base","broom","bsicons","bslib","ca","cachem","car","carData","cli","codetools","colorspace","colourpicker","compiler","data.table","datasets","dendextend","digest","dplyr","DT","ellipsis","evaluate","fansi","fastmap","flexdashboard","forcats","foreach","generics","ggplot2","ggpubr","ggsignif","glue","graphics","grDevices","grid","gridExtra","gtable","heatmaply","hms","htmltools","htmlwidgets","httpuv","httr","iterators","jquerylib","jsonlite","knitr","later","lazyeval","lifecycle","lubridate","magrittr","methods","mime","miniUI","munsell","pillar","pkgconfig","plotly","plyr","processx","promises","ps","purrr","quarto","R6","RColorBrewer","Rcpp","reactable","reactablefmtr","readr","registry","reshape2","rlang","rmarkdown","rstatix","rstudioapi","sass","scales","seriation","shiny","shiny.blueprint","shiny.fluent","shiny.react","shinycssloaders","shinyjqui","shinyjs","shinyWidgets","sparkline","stats","stringi","stringr","tibble","tidyr","tidyselect","tidyverse","timechange","tippy","tools","TSP","tzdb","utf8","utils","vctrs","viridis","viridisLite","webshot","withr","xfun","xtable","yaml"]},{"type":"character","attributes":{},"value":["1.4-5","0.2.1","1.4.1","4.3.0","1.0.5","0.1.2","0.6.1","0.71.1","1.0.8","3.1-2","3.0-5","3.6.1","0.2-19","2.1-0","1.3.0","4.3.0","1.14.8","4.3.0","1.17.1","0.6.31","1.1.3","0.27","0.3.2","0.21","1.0.4","1.1.1","0.6.2","1.0.0","1.5.2","0.1.3","3.4.3","0.6.0","0.6.4","1.6.2","4.3.0","4.3.0","4.3.0","2.3","0.3.3","1.4.2","1.1.3","0.5.7","1.6.2","1.6.11","1.4.6","1.0.14","0.1.4","1.8.4","1.42","1.3.1","0.2.2","1.0.3","1.9.2","2.0.3","4.3.0","0.12","0.1.1.1","0.5.0","1.9.0","2.0.3","4.10.1","1.8.8","3.8.1","1.2.0.1","1.7.5","1.0.2","1.3","2.5.1","1.1-3","1.0.10","0.4.4.9000","2.0.0","2.1.4","0.5-1","1.4.4","1.1.1","2.21","0.7.2","0.14","0.4.6","1.2.1","1.4.2","1.8.0","0.2.0","0.3.0","0.3.0","1.0.0.9011","0.4.1","2.1.0","0.8.0","2.0","4.3.0","1.7.12","1.5.0","3.2.1","1.3.0","1.2.0","2.0.0","0.2.0","0.1.0","4.3.0","1.2-4","0.4.0","1.2.3","4.3.0","0.6.3","0.6.3","0.4.2","0.5.4","2.5.0","0.39","1.8-4","2.3.7"]}]}]}
</script>
<!--/html_preserve-->


<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /container fluid -->

<script type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (_event) {
  if (window.bslib.Card) {
    window.bslib.Card.initializeAllCards();
  }
}); 
</script>
  



</body></html>